pkg/controller/direct/kms/keyhandle/keyhandle_controller.go

```diff
@@ -0,0 +1,199 @@
+	if err != nil {
+		return nil, fmt.Errorf("building KeyHandle client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.KMSKeyHandle{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewKMSKeyHandleRef(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *krm.KMSKeyHandleRef
+	gcpClient *gcp.AutokeyClient
+	desired   *krm.KMSKeyHandle
+	actual    *kmspb.KeyHandle
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting KeyHandle", "name", a.id.External)
+
+	req := &kmspb.GetKeyHandleRequest{Name: a.id.External}
+	keyhandlepb, err := a.gcpClient.GetKeyHandle(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting KeyHandle %q: %w", a.id.External, err)
+	}
+
+	a.actual = keyhandlepb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating KeyHandle", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := KMSKeyHandleSpec_ToProto(mapCtx, &desired.Spec, a.id)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	parent, err := a.id.Parent()
+	if err != nil {
+		return err
+	}
+	req := &kmspb.CreateKeyHandleRequest{
+		Parent:    parent.String(),
+		KeyHandle: resource,
+	}
+	op, err := a.gcpClient.CreateKeyHandle(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating KeyHandle %s: %w", a.id.External, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("KeyHandle %s waiting creation: %w", a.id.External, err)
+	}
+	log.V(2).Info("successfully created KeyHandle", "name", a.id.External)
+
+	status := &krm.KMSKeyHandleStatus{}
+	status.ObservedState = KMSKeyHandleStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = &a.id.External
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update operation not supported for KeyHandle.
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	return fmt.Errorf("update operation not supported for KeyHandle resource")
```

Comment:
This is a KCC update, so it is always called if the `Find` returns the handler object. I think we want to compare the `spec` and the `status` to make sure no field change (return errors if there's a chagne).